File: js/basics/2e_design_ll.js
var Node = function(val, next) {
    this.val = (val === undefined ? 0 : val);
    this.next = (next === undefined ? null : next);
}

var MyLinkedList = function() {
    this.head = null; // Actual head
    this.size = 0;    // Track size
};

/** 
 * @param {number} index
 * @return {number}
 */
MyLinkedList.prototype.get = function(index) {
    if (index < 0 || index >= this.size) return -1;
    let temp = this.head;
    while (index-- > 0) {
        temp = temp.next;
    }
    return temp.val;
};

/** 
 * @param {number} val
 * @return {void}
 */
MyLinkedList.prototype.addAtHead = function(val) {
    let newNode = new Node(val, this.head);
    this.head = newNode;
    this.size++;
};

/** 
 * @param {number} val
 * @return {void}
 */
MyLinkedList.prototype.addAtTail = function(val) {
    let newNode = new Node(val);
    if (!this.head) {
        this.head = newNode;
    } else {
        let temp = this.head;
        while (temp.next) {
            temp = temp.next;
        }
        temp.next = newNode;
    }
    this.size++;
};

/** 
 * @param {number} index 
 * @param {number} val
 * @return {void}
 */
MyLinkedList.prototype.addAtIndex = function(index, val) {
    if (index < 0 || index > this.size) return;
    if (index === 0) return this.addAtHead(val);
    let temp = this.head;
    while (--index > 0) {
        temp = temp.next;
    }
    temp.next = new Node(val, temp.next);
    this.size++;
};

/** 
 * @param {number} index
 * @return {void}
 */
MyLinkedList.prototype.deleteAtIndex = function(index) {
    if (index < 0 || index >= this.size) return;
    if (index === 0) {
        this.head = this.head.next;
    } else {
        let temp = this.head;
        while (--index > 0) {
            temp = temp.next;
        }
        temp.next = temp.next.next;
    }
    this.size--;
};

File: c++/basics/2e_design_ll.cpp
class Node{
    public:
        int val;
        Node* next;
        Node* prev;

        //constructor
        Node(int val){
            this->val = val;
            this->prev = NULL;
            this->next = NULL;
        }
};

class MyLinkedList {
public:
    Node* left;
    Node* right;

    MyLinkedList() {
        left = new Node(0);
        right = new Node(0);
        left->next = right;
        right->prev = left;
    }
    
    int get(int index) {
        if(index<0) return -1;
        Node* temp = left->next;
        while(temp && index){
            temp = temp->next;
            index--;
        }    
        if(temp && temp!= right) return temp->val;
        return -1; 
    }
    
    void addAtHead(int val) {
        Node* newNode = new Node(val);
        newNode->next = left->next;
        left->next->prev = newNode;
        left->next = newNode;
        newNode->prev = left;
    }
    
    void addAtTail(int val) {
        Node* newNode = new Node(val);
        right->prev->next = newNode;
        newNode->prev =right->prev;
        newNode->next = right;
        right->prev = newNode;
    }
    
    void addAtIndex(int index, int val) {
        if(index < 0) return;
        if(index == 0){ 
            addAtHead(val);
            return;
        }
        Node* temp = left;
        while(temp && index>0){
            temp = temp->next;
            index--;
        }
        if(!temp || temp==right || index>0) return;
        Node* fwd = temp->next;
        Node* newNode = new Node(val);
        temp->next = newNode;
        newNode->prev = temp;
        newNode->next = fwd;
        fwd->prev = newNode;
    }
    
    void deleteAtIndex(int index) {
        if(index<0) return;
        Node* temp = left->next;
        while(temp!= right && index>0){
            temp = temp->next;
            index--;
        } 
        Node* back = temp->prev;
        Node* fwd = temp->next;
        if(!fwd || !back) return;
        back->next = fwd;
        fwd->prev = back;
        delete temp;
    }
};

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * MyLinkedList* obj = new MyLinkedList();
 * int param_1 = obj->get(index);
 * obj->addAtHead(val);
 * obj->addAtTail(val);
 * obj->addAtIndex(index,val);
 * obj->deleteAtIndex(index);
 */